rotkehlchen/history/price.py

```diff
@@ -101,61 +160,83 @@ def query_historical_price(from_asset: Asset, to_asset: Asset, timestamp: Timest
             timestamp: The timestamp at which to query the price
 
         May raise:
-        - PriceQueryUnsupportedAsset if from/to asset is missing from price oracles
         - NoPriceForGivenTimestamp if we can't find a price for the asset in the given
         timestamp from the external service.
-        - RemoteError if there is a problem reaching the price oracle server
-        or with reading the response returned by the server
         """
         log.debug(
             'Querying historical price',
             from_asset=from_asset,
             to_asset=to_asset,
             timestamp=timestamp,
         )
-
         if from_asset == to_asset:
             return Price(FVal('1'))
 
-        if from_asset.is_fiat() and to_asset.is_fiat():
-            # if we are querying historical forex data then try something other than cryptocompare
+        is_from_fiat_to_fiat = from_asset.is_fiat() and to_asset.is_fiat()
+        # Querying historical forex data is attempted first via exchangerates API,
+        # and then via any price oracle that has fiat to fiat.
+        if is_from_fiat_to_fiat:
             price = Inquirer().query_historical_fiat_exchange_rates(
                 from_fiat_currency=from_asset,
                 to_fiat_currency=to_asset,
                 timestamp=timestamp,
             )
             if price is not None:
                 return price
-            # else cryptocompare also has historical fiat to fiat data
 
         instance = PriceHistorian()
-        price = None
-        if Inquirer()._cryptocompare.rate_limited_in_last() is False:
+        price = Price(ZERO)
+        for oracle, oracle_instance in instance._oracle_instances:
+            oracle_properties = oracle.properties()
+
+            if is_from_fiat_to_fiat and not oracle_properties.has_fiat_to_fiat:
+                log.debug(
+                    f"Historical price oracle {oracle} can't be used "
+                    f"for fiat to fiat. Skipping",
+                )
+                continue
+
+            if oracle_instance.rate_limited_in_last() is True:
+                log.debug(
+                    f"Historical price oracle {oracle} can't be used "
+                    f"due to rate limits. Skipping",
+                )
+                continue
+
             try:
-                price = instance._cryptocompare.query_historical_price(
+                price = oracle_instance.query_historical_price(
```

Comment:
Standardised method for querying a historical price via historical price oracles.
It is backed by a test that checks all historical price oracles implement this method.